File: frontend/src/modelsConfig.ts
/**
 * Models Configuration - Centralized model definitions for I.R.I.S. Frontend
 */

export interface Resolution {
  width: number
  height: number
}

export interface ModelConfig {
  key: string
  id: string
  description: string
  tier: string
  vram_gb: number
  architecture: string
  pipeline: string
  recommended_steps: number
  recommended_cfg: number
  default_resolution: Resolution
  supports_negative_prompt: boolean
  prompt_style: string
  token_limit: number | null
  lora_supported: boolean
  hf_url: string
  enabled: boolean
}

// Extended model info for UI display
export interface ModelInfo extends ModelConfig {
  best_for: string
  quality_stars: number
  speed_stars: number
  notes: string
}

export interface ModelsConfigData {
  models: Record<string, ModelInfo>
  metadata: {
    version: string
    last_updated: string
    tiers: Record<string, string>
    notes: Record<string, string>
  }
}

// Default model configurations (fallback if API unavailable)
export const defaultModelsConfig: ModelsConfigData = {
  models: {
    'cagliostrolab_animagine-xl-4.0': {
      key: 'cagliostrolab_animagine-xl-4.0',
      id: 'cagliostrolab/animagine-xl-4.0',
      description: 'Animagine XL 4.0',
      tier: 'S',
      vram_gb: 8,
      architecture: 'SDXL',
      pipeline: 'StableDiffusionXLPipeline',
      recommended_steps: 35,
      recommended_cfg: 7.5,
      default_resolution: { width: 1024, height: 1024 },
      supports_negative_prompt: true,
      prompt_style: 'Danbooru tags',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/cagliostrolab/animagine-xl-4.0',
      enabled: true,
      best_for: 'High-quality anime art, SDXL-based',
      quality_stars: 5,
      speed_stars: 3,
      notes: 'Auto-detected model'
    },
    'ojimi_anime-kawai-diffusion': {
      key: 'ojimi_anime-kawai-diffusion',
      id: 'Ojimi/anime-kawai-diffusion',
      description: 'Anime Kawaii Diffusion',
      tier: 'A',
      vram_gb: 2,
      architecture: 'SD 1.5',
      pipeline: 'StableDiffusionPipeline',
      recommended_steps: 28,
      recommended_cfg: 7.5,
      default_resolution: { width: 512, height: 768 },
      supports_negative_prompt: true,
      prompt_style: 'Danbooru tags',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/Ojimi/anime-kawai-diffusion',
      enabled: true,
      best_for: 'Cute anime style, kawaii characters',
      quality_stars: 4,
      speed_stars: 5,
      notes: 'Lightweight SD 1.5 fine-tune for kawaii anime art. Requires ~2GB VRAM.'
    },
    'lykon_dreamshaper-8': {
      key: 'lykon_dreamshaper-8',
      id: 'Lykon/dreamshaper-8',
      description: 'DreamShaper 8',
      tier: 'S',
      vram_gb: 3.5,
      architecture: 'SD 1.5',
      pipeline: 'StableDiffusionPipeline',
      recommended_steps: 30,
      recommended_cfg: 7.5,
      default_resolution: { width: 512, height: 768 },
      supports_negative_prompt: true,
      prompt_style: 'Natural language or Danbooru tags',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/Lykon/dreamshaper-8',
      enabled: true,
      best_for: 'Everything - anime, realistic, fantasy',
      quality_stars: 5,
      speed_stars: 4,
      notes: 'Auto-detected model'
    },
    'onomaai_illustrious-xl': {
      key: 'onomaai_illustrious-xl',
      id: 'OnomaAIResearch/Illustrious-xl-early-release-v0',
      description: 'Illustrious XL v1.0',
      tier: 'S',
      vram_gb: 7,
      architecture: 'SDXL',
      pipeline: 'StableDiffusionXLPipeline',
      recommended_steps: 28,
      recommended_cfg: 7.0,
      default_resolution: { width: 1024, height: 1024 },
      supports_negative_prompt: true,
      prompt_style: 'Danbooru tags',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/OnomaAIResearch/Illustrious-xl-early-release-v0',
      enabled: true,
      best_for: 'High-quality anime/illustration SDXL',
      quality_stars: 5,
      speed_stars: 3,
      notes: 'High-quality anime/illustration SDXL model by OnomaAI. Trained on Danbooru2023. Supports up to 1536x1536 natively.'
    },
    'astraliteheart_pony-diffusion-v6-xl': {
      key: 'astraliteheart_pony-diffusion-v6-xl',
      id: 'AstraliteHeart/pony-diffusion-v6-xl',
      description: 'Pony Diffusion V6 XL',
      tier: 'S',
      vram_gb: 7,
      architecture: 'SDXL',
      pipeline: 'StableDiffusionXLPipeline',
      recommended_steps: 25,
      recommended_cfg: 7.0,
      default_resolution: { width: 1024, height: 1024 },
      supports_negative_prompt: true,
      prompt_style: 'Score tags (score_9, score_8_up)',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/AstraliteHeart/pony-diffusion-v6-xl',
      enabled: true,
      best_for: 'Versatile SDXL anime/character model',
      quality_stars: 5,
      speed_stars: 4,
      notes: 'Versatile SDXL anime/character model. Requires score tags at prompt start. Large LoRA community.'
    },
    'anima_anima-2b': {
      key: 'anima_anima-2b',
      id: 'AnimaLab/anima',
      description: 'Anima 2B',
      tier: 'A',
      vram_gb: 4,
      architecture: 'Transformer (Cosmos-based)',
      pipeline: 'DiffusionPipeline',
      recommended_steps: 20,
      recommended_cfg: 5.0,
      default_resolution: { width: 768, height: 1024 },
      supports_negative_prompt: false,
      prompt_style: 'Natural language',
      token_limit: null,
      lora_supported: false,
      hf_url: 'https://huggingface.co/AnimaLab/anima',
      enabled: true,
      best_for: 'Lightweight transformer-based anime model',
      quality_stars: 4,
      speed_stars: 5,
      notes: 'Lightweight 2B transformer-based anime model. No UNet. No negative prompt support. Fast inference.'
    },
    'hakurei_waifu-diffusion': {
      key: 'hakurei_waifu-diffusion',
      id: 'hakurei/waifu-diffusion',
      description: 'Waifu Diffusion',
      tier: 'B',
      vram_gb: 3.5,
      architecture: 'SD 1.5',
      pipeline: 'StableDiffusionPipeline',
      recommended_steps: 28,
      recommended_cfg: 7.5,
      default_resolution: { width: 512, height: 768 },
      supports_negative_prompt: true,
      prompt_style: 'Danbooru tags',
      token_limit: 77,
      lora_supported: true,
      hf_url: 'https://huggingface.co/hakurei/waifu-diffusion',
      enabled: true,
      best_for: 'Classic anime style, waifus',
      quality_stars: 3,
      speed_stars: 5,
      notes: 'Auto-detected model'
    }
  },
  metadata: {
    version: '2.0.0',
    last_updated: '2026-06-10',
    tiers: {
      'S': 'Best Quality & Versatility',
      'A': 'High Quality',
      'B': 'Good Quality, Fast',
      'C': 'Experimental'
    },
    notes: {
      vram_requirements: 'Approximate VRAM at 512x768 resolution',
      sdxl_warning: 'SDXL models (>8GB VRAM) may cause instability on 16GB systems',
      optimization_note: 'All models support memory optimizations (VAE slicing/tiling, CPU offload)'
    }
  }
}

// Helper functions
export function getModelConfig(modelKey: string): ModelInfo | undefined {
  return defaultModelsConfig.models[modelKey]
}

export function getAllModels(): ModelInfo[] {
  return Object.values(defaultModelsConfig.models).filter(m => m.enabled)
}

export function getModelsByType(type: 'sd15' | 'sdxl' | 'transformer'): ModelInfo[] {
  const typeMap: Record<string, string> = {
    'sd15': 'SD 1.5',
    'sdxl': 'SDXL',
    'transformer': 'Transformer'
  }
  return getAllModels().filter(m => m.architecture === typeMap[type])
}

export function supportsNegativePrompt(modelKey: string): boolean {
  const model = getModelConfig(modelKey)
  return model?.supports_negative_prompt ?? true
}

export function getDefaultResolution(modelKey: string): Resolution {
  const model = getModelConfig(modelKey)
  return model?.default_resolution ?? { width: 512, height: 768 }
}

export function getDefaultSteps(modelKey: string): number {
  const model = getModelConfig(modelKey)
  return model?.recommended_steps ?? 28
}

export function getDefaultCfg(modelKey: string): number {
  const model = getModelConfig(modelKey)
  return model?.recommended_cfg ?? 7.5
}

export function getMaxResolution(architecture: string): Resolution {
  switch (architecture) {
    case 'SD 1.5':
      return { width: 512, height: 768 }
    case 'SDXL':
      return { width: 1536, height: 1536 }
    case 'Transformer (Cosmos-based)':
      return { width: 1024, height: 1024 }
    default:
      return { width: 1024, height: 1024 }
  }
}

export function getVramRequirement(architecture: string): string {
  switch (architecture) {
    case 'SD 1.5':
      return '~2-4GB VRAM'
    case 'SDXL':
      return '~6-8GB VRAM'
    case 'Transformer (Cosmos-based)':
      return '~4GB VRAM'
    default:
      return '~4GB VRAM'
  }
}

export function getTokenLimitInfo(modelKey: string): string {
  const model = getModelConfig(modelKey)
  if (!model) return '77 tokens'
  if (model.token_limit === null) return 'Unlimited'
  return `${model.token_limit} tokens`
}

File: src/api/services/models_config.py
"""
Models Configuration - Centralized model definitions for I.R.I.S.
"""
import json
from pathlib import Path
from typing import Dict, List, Optional, Any
from dataclasses import dataclass


@dataclass
class ModelConfig:
    """Model configuration dataclass"""
    id: str
    description: str
    tier: str
    vram_gb: float
    recommended_steps: int
    recommended_cfg: float
    supports_negative_prompt: bool
    model_type: str
    architecture: str
    pipeline: str
    default_resolution: Dict[str, int]
    prompt_style: str
    token_limit: Optional[int]
    lora_supported: bool
    hf_url: str
    enabled: bool = True
    
    @classmethod
    def from_dict(cls, key: str, data: Dict[str, Any]) -> 'ModelConfig':
        """Create ModelConfig from dictionary"""
        return cls(
            id=data.get('id', ''),
            description=data.get('description', ''),
            tier=data.get('tier', 'B'),
            vram_gb=data.get('vram_gb', 3.5),
            recommended_steps=data.get('recommended_steps', 28),
            recommended_cfg=data.get('recommended_cfg', 7.5),
            supports_negative_prompt=data.get('supports_negative_prompt', True),
            model_type=data.get('model_type', 'sd15'),
            architecture=data.get('architecture', 'SD 1.5'),
            pipeline=data.get('pipeline', 'StableDiffusionPipeline'),
            default_resolution=data.get('default_resolution', {'width': 512, 'height': 768}),
            prompt_style=data.get('prompt_style', 'Natural language'),
            token_limit=data.get('token_limit', 77),
            lora_supported=data.get('lora_supported', True),
            hf_url=data.get('hf_url', ''),
            enabled=not data.get('disabled', False)
        )


class ModelsConfig:
    """Manages model configurations"""
    
    def __init__(self):
        self._configs: Dict[str, ModelConfig] = {}
        self._load_configs()
    
    def _load_configs(self):
        """Load model configurations from JSON file"""
        config_path = Path(__file__).parent.parent.parent.parent / "static" / "config" / "models.json"
        
        try:
            with open(config_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            models = data.get('models', {})
            
            for key, config_data in models.items():
                if not key.startswith('_') and not config_data.get('disabled', False):
                    self._configs[key] = ModelConfig.from_dict(key, config_data)
                    
        except Exception as e:
            print(f"Error loading model configs: {e}")
            # Fallback to empty configs
    
    def get(self, model_key: str) -> Optional[ModelConfig]:
        """Get model config by key"""
        return self._configs.get(model_key)
    
    def get_all(self) -> Dict[str, ModelConfig]:
        """Get all model configurations"""
        return self._configs.copy()
    
    def get_enabled(self) -> List[ModelConfig]:
        """Get all enabled model configurations"""
        return [cfg for cfg in self._configs.values() if cfg.enabled]
    
    def get_by_model_type(self, model_type: str) -> List[ModelConfig]:
        """Get models by type (sd15, sdxl, transformer)"""
        return [cfg for cfg in self._configs.values() 
                if cfg.model_type == model_type and cfg.enabled]
    
    def supports_negative_prompt(self, model_key: str) -> bool:
        """Check if model supports negative prompts"""
        config = self.get(model_key)
        return config.supports_negative_prompt if config else False
    
    def get_default_resolution(self, model_key: str) -> Dict[str, int]:
        """Get default resolution for a model"""
        config = self.get(model_key)
        return config.default_resolution if config else {'width': 512, 'height': 768}
    
    def get_default_steps(self, model_key: str) -> int:
        """Get default steps for a model"""
        config = self.get(model_key)
        return config.recommended_steps if config else 28
    
    def get_default_cfg(self, model_key: str) -> float:
        """Get default CFG scale for a model"""
        config = self.get(model_key)
        return config.recommended_cfg if config else 7.5
    
    def get_pipeline(self, model_key: str) -> str:
        """Get pipeline class name for a model"""
        config = self.get(model_key)
        return config.pipeline if config else 'StableDiffusionPipeline'
    
    def get_info(self, model_key: str) -> Optional[Dict[str, Any]]:
        """Get full model info for frontend display"""
        config = self.get(model_key)
        if not config:
            return None
        
        return {
            'key': model_key,
            'id': config.id,
            'description': config.description,
            'tier': config.tier,
            'vram_gb': config.vram_gb,
            'architecture': config.architecture,
            'pipeline': config.pipeline,
            'recommended_steps': config.recommended_steps,
            'recommended_cfg': config.recommended_cfg,
            'default_resolution': config.default_resolution,
            'supports_negative_prompt': config.supports_negative_prompt,
            'prompt_style': config.prompt_style,
            'token_limit': config.token_limit,
            'lora_supported': config.lora_supported,
            'hf_url': config.hf_url,
            'enabled': config.enabled
        }


# Singleton instance
models_config = ModelsConfig()


def get_model_config(model_key: str) -> Optional[ModelConfig]:
    """Helper function to get model config"""
    return models_config.get(model_key)


def get_all_models() -> Dict[str, ModelConfig]:
    """Helper function to get all models"""
    return models_config.get_all()


def get_model_info(model_key: str) -> Optional[Dict[str, Any]]:
    """Helper function to get model info for frontend"""
    return models_config.get_info(model_key)


def check_negative_prompt_support(model_key: str, negative_prompt: str) -> tuple[bool, str]:
    """
    Check if negative prompt can be used with the model.
    
    Returns:
        tuple: (can_use_negative_prompt, warning_message)
    """
    if not negative_prompt:
        return False, ""
    
    config = models_config.get(model_key)
    if not config:
        return True, ""
    
    if not config.supports_negative_prompt:
        return False, f"Model {config.description} does not support negative prompts. Ignoring negative_prompt."
    
    return True, ""